src/supernova/mod.rs

```diff
@@ -176,92 +176,182 @@ where
   G2: Group<Base = <G1 as Group>::Scalar>,
-  C1: StepCircuit<G1::Scalar>,
-  C2: StepCircuit<G2::Scalar>,
+  C1: EnforcingStepCircuit<G1::Scalar>,
 {
-  fn new(mut claims: Vec<RunningClaim<G1, G2, C1, C2>>) -> Self {
-    let running_claim_params = claims
+  /// new running claim params
+  pub fn new<NC: NonUniformCircuit<G1, G2, C1>>(
+    non_unifrom_circuit: &NC,
+    // optfn1: Option<CommitmentKeyHint<G1>>,
+    // optfn2: Option<CommitmentKeyHint<G2>>,
+  ) -> Self {
+    // uhhh if this aint zero this doesn't work
+    let _initial_pc = non_unifrom_circuit.initial_program_counter();
+    let num_circuits = non_unifrom_circuit.num_circuits();
+
+    let c_primarys = (0..num_circuits)
+      .map(|i| non_unifrom_circuit.primary_circuit(i))
+      .collect::<Vec<_>>();
+    let c_secondary = non_unifrom_circuit.secondary_circuit();
+
+    let pp_vec = c_primarys
       .iter()
-      .map(|c| c.get_public_params())
+      .map(|c_primary| PublicParams::setup_without_commitkey(c_primary, &c_secondary, num_circuits))
       .collect::<Vec<_>>();
 
-    let (ck_primary, ck_secondary) = compute_commitment_keys(&running_claim_params);
+    let mut running_claim_params = RunningClaimParams {
+      pp_vec,
+      digest: OnceCell::new(),
+      stage: Stage::Incomplete,
+      _p: PhantomData,
+    };
+
+    running_claim_params.complete();
+    running_claim_params
+  }
+
+  /// Create a [RunningClaimParams] from a vector of raw [PublicParams].
+  /// We check the state of the commitment keys of the given params and
+  /// realign them if necessary. If a digest is given, we assume that it
+  /// will correctly match the public params after realigning them, and
+  /// we do not check for its validity.
+  pub fn from_pp_vec(pp_vec: Vec<PublicParams<G1, G2>>, digest: OnceCell<G1::Scalar>) -> Self {
+    let num_pp = pp_vec.len();
+    let primary_len = &pp_vec[0].ck_primary.as_ref().map_or(0, Len::length);
+    let secondary_len = &pp_vec[0].ck_secondary.as_ref().map_or(0, Len::length);
+    let num_complete = pp_vec
+      .iter()
+      .filter(|pp| {
+        pp.ck_primary.is_some()
+          && pp.ck_secondary.is_some()
+          && &pp.ck_primary.as_ref().unwrap().length() == primary_len
+          && &pp.ck_secondary.as_ref().unwrap().length() == secondary_len
+      })
+      .count();
+
+    let stage = if num_complete == num_pp {
+      Stage::Complete
+    } else {
+      Stage::Incomplete
+    };
+
+    let mut running_claim_params = RunningClaimParams {
+      pp_vec,
+      digest,
+      stage,
+      _p: PhantomData,
+    };
+
+    running_claim_params.complete();
+    running_claim_params
+  }
 
-    for claim in &mut claims {
-      claim.set_commitment_key(ck_primary.clone(), ck_secondary.clone());
+  /// Compute primary and secondary commitment keys sized to handle the largest of the circuits in the provided
+  /// `PublicParams`.
+  fn compute_commitment_keys(&self) -> (CommitmentKey<G1>, CommitmentKey<G2>) {
```

Comment:
This is where the optfn1, optfn2 parameters from above come in. They need to be parametrized by a preference expressed as a static function of the SNARK that will eventually be used to compress the proof. See #5. We should open an issue to fix this across SuperNova.

Reply:
Done in #53